controller/source/rocksdb_server/server.cpp

```diff
@@ -0,0 +1,95 @@
+#include <iostream>
+#include <string>
+#include <span>
+#include <vector>
+#include <memory>
+#include <boost/asio.hpp>
+
+#include "rocksdb_proxy.hpp"
+
+using boost::asio::ip::tcp;
+
+class session : public std::enable_shared_from_this<session> {
+public:
+  session(tcp::socket socket, std::shared_ptr<rocksdb_proxy> rocksdb_proxy)
+      : m_socket(std::move(socket))
+      , m_rocksdb_proxy(std::move(rocksdb_proxy))
+  {
+  }
+
+  void start() {
+    do_read();
+  }
+
+private:
+  void do_read() {
+    auto self(shared_from_this());
+    boost::asio::async_read_until(m_socket, m_buffer, '\n',
+      [this, self](boost::system::error_code error_code, std::size_t length) {
+      if (!error_code) {
+        // put buffered data into string excluding the last new line character
+        std::string raw_query(boost::asio::buffers_begin(m_buffer.data()), boost::asio::buffers_end(m_buffer.data()) - 1);
+        m_buffer.consume(length);
+        query_message query = query_message::deserialize(raw_query);
+        response_message response = m_rocksdb_proxy->execute(query);
+        std::string raw_response = response.serialize();
+        boost::asio::write(m_socket, boost::asio::buffer(raw_response));
+        do_read();
```

Comment:
Couldn't this recursion be done with a while loop? (it's fine.. just to avoid the warning)